indexing-service/src/main/java/org/apache/druid/indexing/common/task/AppendTask.java

```diff
@@ -0,0 +1,336 @@
+import org.apache.druid.timeline.partition.NumberedPartialShardSpec;
+import org.joda.time.DateTime;
+import org.joda.time.Interval;
+
+import javax.annotation.Nonnull;
+import java.io.IOException;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.function.Function;
+
+/**
+ * A test Task which mimics an appending task by having similar interactions with the overlord.
+ *
+ * Begins running by acquiring an APPEND lock and immediately allocates pending segments.
+ *
+ * Task ends after publishing these pending segments and relevant metadata entries in a transaction
+ * Replace lock exists with version V3
+ * V0 -> PS0 -> append task begins
+ * V1 -> (S1-0) -> replace task begins and published and completed
+ * V2 -> (S2-0) -> replace task begins and published and completed
+ * V3 -> Replace acquired lock
+ *
+ * append task publishes PS0 -> S0-0, S1-1, S2-1 ; (S0-0, V3); Append task has completed
+ *
+ *
+ * (S3-0, 1) today
+ * (S3-0, 2), (S3-1, 2) needs to happen
+ * V3 replace task finishes -> (S3-0, S0-0 == S3-1)
+ *
+ * segment metadata -> Publish all pending segments and also create copies for greater versions for which used segments exist
+ * forward metadata -> Publish mapping of the original segments to the EXCLUSIVE lock held for the same interval when present
+ */
+public class AppendTask extends AbstractTask
+{
+  private final Interval interval;
+  private final Granularity segmentGranularity;
+  private final String lockType;
+  private final int priority;
+  private final int numPartitions;
+  private final CountDownLatch readyLatch = new CountDownLatch(1);
+  private final CountDownLatch runLatch = new CountDownLatch(1);
+  private final CountDownLatch segmentAllocationComplete = new CountDownLatch(1);
+  private final CountDownLatch runComplete = new CountDownLatch(1);
+  private final CountDownLatch readyComplete = new CountDownLatch(1);
+
+  private final Set<SegmentIdWithShardSpec> pendingSegments = new HashSet<>();
+
+  private TaskToolbox toolbox;
+  private final AtomicInteger sequenceId = new AtomicInteger(0);
+
+  public AppendTask(
+      @JsonProperty("id") String id,
+      @JsonProperty("dataSource") String dataSource,
+      @JsonProperty("interval") Interval interval,
+      @JsonProperty("segmentGranularity") Granularity segmentGranularity,
+      @JsonProperty("context") Map<String, Object> context
+  )
+  {
+    super(
+        id == null ? StringUtils.format("replace_%s_%s", DateTimes.nowUtc(), UUID.randomUUID().toString()) : id,
+        dataSource == null ? "none" : dataSource,
+        context,
+        IngestionMode.APPEND
+    );
+    this.interval = interval;
+    this.segmentGranularity = segmentGranularity;
+    this.lockType = getContextValue(Tasks.TASK_LOCK_TYPE, "EXCLUSIVE");
```

Comment:
If this is an Append task, why would it use any lock other than the append lock?

Reply:
Fixed